src/main/scala/io/iohk/ethereum/blockchain/sync/regular/StateNodeFetcher.scala

```diff
@@ -0,0 +1,136 @@
+
+  override def onMessage(message: StateNodeFetcherCommand): Behavior[StateNodeFetcherCommand] = {
+    message match {
+      case StateNodeFetcher.FetchStateNode(hash, sender) =>
+        log.debug("Start fetching state node")
+        requestStateNode(hash)
+        requester = Some(StateNodeRequester(hash, sender))
+        Behaviors.same
+      case AdaptedMessage(peer, NodeData(values)) if requester.isDefined =>
+        log.debug("Received state node response from peer {}", peer)
+
+        requester
+          .collect(stateNodeRequester => {
+            val validatedNode = values
+              .asRight[String]
+              .ensure(s"Empty response from peer $peer, blacklisting")(_.nonEmpty)
+              .ensure("Fetched node state hash doesn't match requested one, blacklisting peer")(nodes =>
+                stateNodeRequester.hash == kec256(nodes.head))
+
+            validatedNode match {
+              case Left(err) =>
+                log.debug(err)
+                peersClient ! BlacklistPeer(peer.id, err)
+                context.self ! StateNodeFetcher.FetchStateNode(stateNodeRequester.hash, stateNodeRequester.replyTo)
+                Behaviors.same[StateNodeFetcherCommand]
+              case Right(node) =>
+                stateNodeRequester.replyTo ! FetchedStateNode(NodeData(node))
+                requester = None
+                Behaviors.same[StateNodeFetcherCommand]
+            }
+          })
+          .getOrElse(Behaviors.same)
+
+      case StateNodeFetcher.RetryStateNodeRequest if requester.isDefined =>
+        log.debug("Something failed on a state node request, trying again")
+        requester
+          .collect(stateNodeRequester =>
+            context.self ! StateNodeFetcher.FetchStateNode(stateNodeRequester.hash, stateNodeRequester.replyTo))
+        Behaviors.same
+      case _ => Behaviors.unhandled
+    }
+  }
+
+  private def requestStateNode(hash: ByteString): Unit = {
+    val resp = makeRequest(Request.create(GetNodeData(List(hash)), BestPeer), StateNodeFetcher.RetryStateNodeRequest)
+    context.pipeToSelf(resp.runToFuture) {
+      case Success(res) => res
+      case Failure(_) => StateNodeFetcher.RetryStateNodeRequest
+    }
+  }
+
+  private def makeRequest(
+      request: Request[_],
+      responseFallback: StateNodeFetcherCommand): Task[StateNodeFetcherCommand] =
+    Task
+      .deferFuture(peersClient ? request)
+      .tap(blacklistPeerOnFailedRequest)
+      .flatMap(handleRequestResult(responseFallback))
+      .onErrorHandle { error =>
+        log.error("Unexpected error while doing a request", error)
+        responseFallback
+      }
+
+  private def blacklistPeerOnFailedRequest(msg: Any): Unit = msg match {
+    case RequestFailed(peer, reason) => peersClient ! BlacklistPeer(peer.id, reason)
+    case _ => ()
+  }
+
+  private def handleRequestResult(fallback: StateNodeFetcherCommand)(msg: Any): Task[StateNodeFetcherCommand] =
```

Comment:
I see that the `HeadersFetcher`, `BodiesFetcher` and `StateNodeFetcher` share some logic and in the `makeRequest` and `handleRequestResult` methods, can these be abstracted into a trait?

Reply:
sorry I forgot to mark this PR as WIP. this was just trying to make it work. But it didn't, we won't proceed with these changes